### PR TITLE
fix near-distance for chained searches in dstu3

### PR DESCRIPTION
**Pull request by fil512**
ready for review

### CODE REVIEW

**Comment by fil512** [comment]
The test classes were getting big, so I pulled the distance tests out into separate classes.

The new tests that were added were the PractitionerRole tests in `ResourceProviderDstu3DistanceTest.java`

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+			List<List<IQueryParameterType>> paramAndList = theParams.get("location");
+			ReferenceParam referenceParam = getChainedLocationNearDistanceParam(paramAndList);
+			if (referenceParam != null) {
+				QuantityParam quantityParam = new QuantityParam(referenceParam.getValue());
+				theParams.setNearDistanceParam(quantityParam);
+			}
+		}
+	}
+
+	private static ReferenceParam getChainedLocationNearDistanceParam(List<List<IQueryParameterType>> theParamAndList) {
```
Comment (tadgh) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: This function is a tad messy. I'm trying to think of ways in which we could simplify this, as its starting to look like arrow code. Am I right in thinking the pseudocode For this essentially:
```
1. Stream over all ParamAndLists and their child ParamOrLists
2. if more than a single Param of type SP_NEAR_DISTANCE  exists, cumulatively, over any of the ParamOrLists, then throw illegal argument exception. 
3. If 0 exist, return null, 
4. If exactly 1 exists, return it, but also remove it from its paramOrList, and if that causes the subsequent parent ParamAndList to be empty, remove it too.
```
Have I got the gist?

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+			throw new IllegalArgumentException("Only one " + ca.uhn.fhir.model.dstu2.resource.Location.SP_NEAR_DISTANCE + " parameter may be present");
+		}
+		List<IQueryParameterType> paramOrList = theParamAndList.get(0);
+		if (paramOrList.isEmpty()) {
+			return null;
+		}
+		if (paramOrList.size() > 1) {
+			throw new IllegalArgumentException("Only one " + ca.uhn.fhir.model.dstu2.resource.Location.SP_NEAR_DISTANCE + " parameter may be present");
+		}
+		return (QuantityParam) paramOrList.get(0);
```
Comment (tadgh) [inline hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Is this functionally equivalent to 
```java
long sum = theParamAndList.stream().mapToLong(Collection::size).sum();
//No Near Distance Param
if (sum  == 0) {
	return null;
// A single Near Distance Param
} else if (sum == 1) {
	return (QuantityParam) theParamAndList.get(0).get(0);
//too many near distance params
} else {
	throw new IllegalArgumentException("Only one " + ca.uhn.fhir.model.dstu2.resource.Location.SP_NEAR_DISTANCE + " parameter may be present");
```		
Or are there some possible cases I'm missing?

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ContainedTest.java [hunk 3]
```diff
@@ -47,12 +51,9 @@ public void testIndexContained() {
...
 		
-		SearchParameterMap map;
-		
-//		map = new SearchParameterMap();
-//		map.add(Observation.SP_CODE, new TokenParam(null, "some observation").setModifier(TokenParamModifier.TEXT));
-//		assertThat(toUnqualifiedVersionlessIdValues(myObservationDao.search(map)), containsInAnyOrder(toValues(id1, id2)));
-
+		SearchParameterMap map = new SearchParameterMap();
+		map.add(Observation.SP_CODE, new TokenParam(null, "some observation").setModifier(TokenParamModifier.TEXT));
+		assertThat(toUnqualifiedVersionlessIdValues(myObservationDao.search(map)), containsInAnyOrder(toValues(oid1, oid2)));
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/dstu3/FhirResourceDaoDstu3ContainedTest.java]: No idea.  I was helping Ruslan with something and came across it.  They were commented out 3 years ago.  I uncommented them, the test passed, so why not?

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+import ca.uhn.fhir.model.api.IQueryParameterType;
+import ca.uhn.fhir.rest.param.QuantityParam;
+import ca.uhn.fhir.rest.param.ReferenceParam;
+import org.hl7.fhir.dstu3.model.Location;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+
+import java.util.List;
+
+public class DistanceHelper {
+	public static void setNearDistance(Class<? extends IBaseResource> theResourceType, SearchParameterMap theParams) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Good suggestion.  Done.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+		if (theResourceType == Location.class && theParams.containsKey(Location.SP_NEAR_DISTANCE)) {
+			List<List<IQueryParameterType>> paramAndList = theParams.get(Location.SP_NEAR_DISTANCE);
+			QuantityParam quantityParam = getNearDistanceParam(paramAndList);
+			theParams.setNearDistanceParam(quantityParam);
+
+			// Need to remove near-distance or it we'll get a hashcode predicate for it
+			theParams.remove(Location.SP_NEAR_DISTANCE);
+		} else if (theParams.containsKey("location")) {
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Completely agree.  I looked around, and could not find a good source for it.  It's really referring to a named path in multiple resources and their shared naming is convention, not structural, so I left it as a string.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+		ReferenceParam retval = null;
+		List<IQueryParameterType> andParamToRemove = null;
+		for (List<IQueryParameterType> paramOrList : theParamAndList) {
+			IQueryParameterType orParamToRemove = null;
+			for (IQueryParameterType param : paramOrList) {
+				if (param instanceof ReferenceParam) {
+					ReferenceParam referenceParam = (ReferenceParam) param;
+					if (Location.SP_NEAR_DISTANCE.equals(referenceParam.getChain())) {
+						if (retval != null) {
+							throw new IllegalArgumentException("Only one " + ca.uhn.fhir.model.dstu2.resource.Location.SP_NEAR_DISTANCE + " parameter may be present");
```
[the remark this answers is not in the document]
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: That was my intention--I could have sworn I fixed that.  Strange.  Fixed again

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java (hunk 1 above)
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Yes!  This yucky method is the reason why I spent an hour trying to do this processing at the point where search parameters are added initially.  I'd love to see a cleaner way to write this if you can come up with a way.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java (hunk 2 above)
Comment (fil512) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Much nicer!  I replaced with your code and re-ran all the tests.

File: hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java (hunk 1 above)
Comment (tadgh) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: I may be able to streamify this stuff, but the fact that we have to remove the offending params post-selection means its probably gonna stay yucky. Let's open an issue to just think about this I think. I'll poke at it when I've got some downtime

Comment (tadgh) [reply hapi-fhir-jpaserver-searchparam/src/main/java/ca/uhn/fhir/jpa/searchparam/util/DistanceHelper.java]: Created #1746 to address this.

**Pull request merged**